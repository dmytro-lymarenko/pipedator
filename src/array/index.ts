import { test } from '../test';

export function array<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => Array.isArray(value), message || 'Value should be an array');
}
