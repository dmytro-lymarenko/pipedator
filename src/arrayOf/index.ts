import { every } from '../every';

/**
 * An alias for {@link every}
 */
export const arrayOf = every;
