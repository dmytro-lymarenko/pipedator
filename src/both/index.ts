/**
 * @module validators
 */

import { pipe } from '../pipe';

export const both = pipe;
