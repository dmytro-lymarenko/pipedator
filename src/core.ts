export interface ValidationError {
	message: string;
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
}

export interface ValidationContext {
	value: any;
	path: string[];
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	pipe: (...validators: Validator[]) => Validator;
}

export function pipe(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			// find the first error in pipe
			let i: number = 0;
			let error: ValidationError | null = null;

			while (i < validators.length && error === null) {
				error = validators[i].validate(value, ctx);
				i = i + 1;
			}

			return error;
		},
	});
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator(options: CreateValidatorOptions): Validator {
	function validate(value: any, ctx?: ValidationContext) {
		return options.validate(value, ctx || { value, path: [] });
	}

	return {
		validate,
		pipe: (...validators) =>
			pipe(
				createValidator({ validate }),
				...validators
			),
	};
}
