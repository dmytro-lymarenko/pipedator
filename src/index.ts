// This file is generated automatically by `scripts/buildIndex.ts`. Please, don't change it.

export * from './abstractShape';
export * from './alt';
export * from './alternative';
export * from './both';
export * from './core';
export * from './either';
export * from './empty';
export * from './equalTo';
export * from './every';
export * from './failure';
export * from './greater';
export * from './ignore';
export * from './keys';
export * from './nullable';
export * from './optional';
export * from './pipe';
export * from './shape';
export * from './some';
export * from './string';
export * from './success';
export * from './ternary';
export * from './test';
export * from './tuple';
export * from './valid';
export * from './values';
export * from './valuesByKeys';
export * from './when';
