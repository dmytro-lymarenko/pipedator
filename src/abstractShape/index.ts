import { createValidator, findFirstError, isValidationError, Validator } from '../core';

export function abstractShape<Key>(
	keys: Key[],
	shape: (key: Key) => Validator,
	options?: { onlyFirstError?: boolean },
	message?: string
) {
	return createValidator({
		validate: (value, ctx) => {
			if (options && options.onlyFirstError) {
				const { error } = findFirstError(i => shape(keys[i]), i => value[keys[i]], keys.length, i => ({
					...ctx,
					path: [...ctx.path, keys[i].toString()],
				}));

				if (error && message) {
					return ctx.generateError({
						value,
						message,
						path: ctx.path,
					});
				}

				return error;
			}

			const errors = keys
				.map(key =>
					shape(key).validate(value[key], {
						...ctx,
						path: [...ctx.path, key.toString()],
					})
				)
				.filter(isValidationError)
				.reduce((res, err) => [...res, ...err.errors], []);

			if (errors.length > 0) {
				if (message) {
					return ctx.generateError({
						value,
						message,
						path: ctx.path,
					});
				}

				return ctx.generateErrors(errors);
			}

			return null;
		},
	});
}
