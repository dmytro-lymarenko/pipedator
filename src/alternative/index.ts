import { createValidator, getFirstErrors, Validator, getCurrentPath } from '../core';

export function alternative<V1, V2>(v1: Validator<V1>, v2: Validator<V2>): Validator<V1 | V2>;
export function alternative<V1, V2, V3>(v1: Validator<V1>, v2: Validator<V2>, v3: Validator<V3>): Validator<V1 | V2 | V3>;
export function alternative<V1, V2, V3, V4>(
	v1: Validator<V1>,
	v2: Validator<V2>,
	v3: Validator<V3>,
	v4: Validator<V4>
): Validator<V1 | V2 | V3 | V4>;
export function alternative<V1, V2, V3, V4, V5>(
	v1: Validator<V1>,
	v2: Validator<V2>,
	v3: Validator<V3>,
	v4: Validator<V4>,
	v5: Validator<V5>
): Validator<V1 | V2 | V3 | V4 | V5>;
export function alternative<V1, V2, V3, V4, V5, V6>(
	v1: Validator<V1>,
	v2: Validator<V2>,
	v3: Validator<V3>,
	v4: Validator<V4>,
	v5: Validator<V5>,
	v6: Validator<V6>
): Validator<V1 | V2 | V3 | V4 | V5 | V6>;
export function alternative<V1, V2, V3, V4, V5, V6, V7>(
	v1: Validator<V1>,
	v2: Validator<V2>,
	v3: Validator<V3>,
	v4: Validator<V4>,
	v5: Validator<V5>,
	v6: Validator<V6>,
	v7: Validator<V7>
): Validator<V1 | V2 | V3 | V4 | V5 | V6 | V7>;
export function alternative<V1, V2, V3, V4, V5, V6, V7, V8>(
	v1: Validator<V1>,
	v2: Validator<V2>,
	v3: Validator<V3>,
	v4: Validator<V4>,
	v5: Validator<V5>,
	v6: Validator<V6>,
	v7: Validator<V7>,
	v8: Validator<V8>
): Validator<V1 | V2 | V3 | V4 | V5 | V6 | V7 | V8>;

export function alternative(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			if (validators.length === 0) {
				return null;
			}
			// find the first success validator
			const errors = getFirstErrors(i => validators[i], () => value, () => ctx, validators.length);

			if (errors.length === validators.length) {
				// here all validators returned an error
				if (errors.length === 1) {
					return errors[0];
				}

				return {
					path: getCurrentPath(ctx),
					message: `Either ${errors.map(error => error.message).join(' OR ')}`,
					children: null,
				};
			}

			return null;
		},
	});
}
