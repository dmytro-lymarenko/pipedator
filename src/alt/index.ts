import { alternative } from '../alternative';

export const alt = alternative;
