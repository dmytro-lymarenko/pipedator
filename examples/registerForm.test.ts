import { shape, pipe, string, greater, less, match, alt, number, optional } from '../src';

/*
const schema = Joi.object().keys({
    username: Joi.string().alphanum().min(3).max(30).required(),
    password: Joi.string().regex(/^[a-zA-Z0-9]{3,30}$/),
    access_token: [Joi.string(), Joi.number()],
    birthyear: Joi.number().integer().min(1900).max(2013),
    email: Joi.string().email({ minDomainSegments: 2 })
}).with('username', 'birthyear').without('password', 'access_token');
*/

const schema = shape({
	username: pipe(
		string(),
		greater(3),
		less(30)
	),
	password: optional(
		pipe(
			string(),
			match(/^[a-zA-Z0-9]{3,30}$/)
		)
	),
	access_token: optional(alt(string(), number())),
	birthyear: optional(
		pipe(
			number(),
			greater(1900),
			less(2013)
		)
	),
	email: optional(string()),
});

describe('Register form test', () => {
	it('should pass the validation', () => {
		expect(
			schema.validate({
				username: 'John',
			})
		).toMatchSnapshot();
	});
});
