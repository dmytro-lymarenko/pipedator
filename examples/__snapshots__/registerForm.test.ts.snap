// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register form test should pass the validation 1`] = `
Object {
  "children": Array [
    Object {
      "children": null,
      "message": "Value should be greater than 3",
      "path": "username",
    },
  ],
  "message": "Value should have a valid shape",
  "path": "",
}
`;
